### PR TITLE
test: updated test coverage for InterchainTokenService.sol

### PR DESCRIPTION
**Pull request by ahramy**
https://axelarnetwork.atlassian.net/browse/AXE-4555

## Before: 
- `Branch` test coverage for ` InterchainTokenService.sol ` was `95.68%`
- `Branch` test coverage for all files was `95.45%`

## After the change
- `Branch` test coverage for ` InterchainTokenService.sol ` is `100%`
- `Branch` test coverage for all files was `97.44%`
```
----------------------------------------|----------|----------|----------|----------|----------------|
File                                    |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
----------------------------------------|----------|----------|----------|----------|----------------|
 contracts/                             |      100 |    98.73 |      100 |    99.18 |                |
  InterchainTokenFactory.sol            |      100 |      100 |      100 |      100 |                |
  InterchainTokenService.sol            |      100 |      100 |      100 |      100 |                |
  TokenHandler.sol                      |      100 |    94.23 |      100 |    95.38 |     78,116,155 |
 contracts/executable/                  |      100 |      100 |      100 |      100 |                |
  InterchainTokenExecutable.sol         |      100 |      100 |      100 |      100 |                |
  InterchainTokenExpressExecutable.sol  |      100 |      100 |      100 |      100 |                |
 contracts/interchain-token/            |      100 |      100 |      100 |      100 |                |
  ERC20.sol                             |      100 |      100 |      100 |      100 |                |
  ERC20Permit.sol                       |      100 |      100 |      100 |      100 |                |
  InterchainToken.sol                   |      100 |      100 |      100 |      100 |                |
  InterchainTokenStandard.sol           |      100 |      100 |      100 |      100 |                |
 contracts/interfaces/                  |      100 |      100 |      100 |      100 |                |
  IAddressTracker.sol                   |      100 |      100 |      100 |      100 |                |
  IBaseTokenManager.sol                 |      100 |      100 |      100 |      100 |                |
  IERC20BurnableFrom.sol                |      100 |      100 |      100 |      100 |                |
  IERC20MintableBurnable.sol            |      100 |      100 |      100 |      100 |                |
  IERC20Named.sol                       |      100 |      100 |      100 |      100 |                |
  IFlowLimit.sol                        |      100 |      100 |      100 |      100 |                |
  IGatewayCaller.sol                    |      100 |      100 |      100 |      100 |                |
  IInterchainToken.sol                  |      100 |      100 |      100 |      100 |                |
  IInterchainTokenDeployer.sol          |      100 |      100 |      100 |      100 |                |
  IInterchainTokenExecutable.sol        |      100 |      100 |      100 |      100 |                |
  IInterchainTokenExpressExecutable.sol |      100 |      100 |      100 |      100 |                |
  IInterchainTokenFactory.sol           |      100 |      100 |      100 |      100 |                |
  IInterchainTokenService.sol           |      100 |      100 |      100 |      100 |                |
  IInterchainTokenStandard.sol          |      100 |      100 |      100 |      100 |                |
  IMinter.sol                           |      100 |      100 |      100 |      100 |                |
  IOperator.sol                         |      100 |      100 |      100 |      100 |                |
  ITokenHandler.sol                     |      100 |      100 |      100 |      100 |                |
  ITokenManager.sol                     |      100 |      100 |      100 |      100 |                |
  ITokenManagerDeployer.sol             |      100 |      100 |      100 |      100 |                |
  ITokenManagerImplementation.sol       |      100 |      100 |      100 |      100 |                |
  ITokenManagerProxy.sol                |      100 |      100 |      100 |      100 |                |
  ITokenManagerType.sol                 |      100 |      100 |      100 |      100 |                |
  ITransmitInterchainToken.sol          |      100 |      100 |      100 |      100 |                |
 contracts/proxies/                     |      100 |      100 |      100 |      100 |                |
  InterchainProxy.sol                   |      100 |      100 |      100 |      100 |                |
  TokenManagerProxy.sol                 |      100 |      100 |      100 |      100 |                |
 contracts/token-manager/               |      100 |    96.15 |      100 |      100 |                |
  TokenManager.sol                      |      100 |    96.15 |      100 |      100 |                |
 contracts/types/                       |      100 |      100 |      100 |      100 |                |
  InterchainTokenServiceTypes.sol       |      100 |      100 |      100 |      100 |                |
 contracts/utils/                       |      100 |     87.5 |      100 |    97.44 |                |
  Create3AddressFixed.sol               |      100 |      100 |      100 |      100 |                |
  Create3Fixed.sol                      |      100 |    83.33 |      100 |      100 |                |
  FlowLimit.sol                         |      100 |      100 |      100 |      100 |                |
  GatewayCaller.sol                     |      100 |    83.33 |      100 |    85.71 |         64,116 |
  InterchainTokenDeployer.sol           |      100 |       75 |      100 |      100 |                |
  Minter.sol                            |      100 |      100 |      100 |      100 |                |
  Operator.sol                          |      100 |      100 |      100 |      100 |                |
  RolesConstants.sol                    |      100 |      100 |      100 |      100 |                |
  TokenManagerDeployer.sol              |      100 |       50 |      100 |      100 |                |
----------------------------------------|----------|----------|----------|----------|----------------|
All files                               |      100 |    97.44 |      100 |    99.11 |                |
----------------------------------------|----------|----------|----------|----------|----------------|
```

### CODE REVIEW

File: test/InterchainTokenService.js [hunk 1]
```diff
@@ -84,6 +87,36 @@ describe('Interchain Token Service', () => {
         return [token, tokenManager, tokenId];
     };
 
+    deployFunctionsTest.lockUnlock = async function deployNewLockUnlock(
```
Comment (ahramy) [inline test/InterchainTokenService.js:90]: Note: There are multiple `async function deployNewLockUnlock` and `async function deployNewGateway` functions defined, need to refactor the code. (Would be a separate PR)

File: test/InterchainTokenService.js [hunk 2]
```diff
@@ -11,13 +11,16 @@ const {
...
-const { deployAll, deployContract, deployInterchainTokenService } = require('../scripts/deploy');
+const { getRandomBytes32, getRandomInt, expectRevert, gasReporter, getEVMVersion, itsHubChainName } = require('./utils');
+const { deployAll, deployContract, deployInterchainTokenService, deployAllWithTestGatewayCaller } = require('../scripts/deploy');
 
 const MESSAGE_TYPE_INTERCHAIN_TRANSFER = 0;
 const MESSAGE_TYPE_DEPLOY_INTERCHAIN_TOKEN = 1;
 const MESSAGE_TYPE_DEPLOY_TOKEN_MANAGER = 2;
-const INVALID_MESSAGE_TYPE = 3;
+/* eslint-disable-next-line no-unused-vars */
+const MESSAGE_TYPE_SEND_TO_HUB = 3;
```
Comment (ahramy) [inline test/InterchainTokenService.js:21]: Note: This part needs to be refactored. e.g. move const variables to `constants.js`

File: test/InterchainTokenService.js (hunk 1 above)
Comment (milapsheth) [reply test/InterchainTokenService.js:91]: I generally prefer to avoid making code quality worse temporarily to fix it later. Instead, we should try adding the test in a cleaner manner even though it might take more work than originally expected for the task.

File: scripts/deploy.js [hunk 3]
```diff
@@ -138,10 +138,53 @@ async function deployAll(
     };
 }
 
+async function deployAllWithTestGatewayCaller(
```
Comment (milapsheth) [inline scripts/deploy.js:141]: Instead of adding a duplicate, change `deployAll` to accept an optional arg (say `contractOverrides`) to override contracts to use instead of deploying it's own. Example shown below

File: scripts/deploy.js [hunk 4]
```diff
@@ -138,10 +138,53 @@ async function deployAll(
...
+    const gateway = await deployMockGateway(wallet);
+    const gasService = await deployGasService(wallet);
+
+    const interchainTokenServiceAddress = await getCreate3Address(create3Deployer.address, wallet, deploymentKey);
+    const tokenManagerDeployer = await deployContract(wallet, 'TokenManagerDeployer', []);
+    const interchainToken = await deployContract(wallet, 'InterchainToken', [interchainTokenServiceAddress]);
+    const interchainTokenDeployer = await deployContract(wallet, 'InterchainTokenDeployer', [interchainToken.address]);
+    const tokenManager = await deployContract(wallet, 'TokenManager', [interchainTokenServiceAddress]);
+    const tokenHandler = await deployContract(wallet, 'TokenHandler', [gateway.address]);
+    const gatewayCaller = await deployContract(wallet, 'TestGatewayCaller');
```
Comment (milapsheth) [inline scripts/deploy.js:160]:
```suggestion
    const gatewayCaller = await deployContract(wallet, contractOverrides.GatewayCaller || 'GatewayCaller');
```
This can optionally overriden by the caller. You can add a docstring on the method that this is useful for mocking contracts in tests

**Comment by milapsheth** [comment]
Looks like Gh actions were cancelled

**Review by ahramy — COMMENTED** [review]
updated code, fixed lint issues and GH action tests.

File: contracts/test/utils/TestGatewayCaller.sol [hunk 5]
```diff
@@ -0,0 +1,10 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+contract TestGatewayCaller {
+    function delegatecall(bytes memory) external pure returns (bool, bytes memory) {
+        // solhint-disable-next-line reason-string
+        revert();
```
Comment (milapsheth) [inline contracts/test/utils/TestGatewayCaller.sol:8]: address this lint. you can define an error code

File: scripts/deploy.js (hunk 4 above)
Comment (milapsheth) [reply scripts/deploy.js:159]: test code shouldn't be added to this folder. I prefer adding a proper deployment fix similar to my suggestion above instead of a custom test deployment method that we now have to maintain. Or, do the custom deployment within the test itself

File: test/InterchainTokenService.js [hunk 6]
```diff
@@ -12,22 +12,25 @@ const {
...
+    MINT_BURN_FROM,
+    LOCK_UNLOCK,
+    LOCK_UNLOCK_FEE_ON_TRANSFER,
+    MINT_BURN,
+    GATEWAY,
+    OPERATOR_ROLE,
+    FLOW_LIMITER_ROLE,
+    ITS_HUB_CHAIN_NAME,
+    ITS_HUB_ROUTING_IDENTIFIER,
+    itsHubAddress,
```
Comment (milapsheth) [inline test/InterchainTokenService.js:32]: switch to capital casing

File: test/InterchainTokenService.js [hunk 7]
```diff
@@ -195,6 +191,9 @@ describe('Interchain Token Service', () => {
             return [token, tokenManager, tokenId];
         };
 
+    deployFunctions.lockUnlock = deployNewLockUnlock;
+    deployFunctions.gateway = deployNewGateway;
+    deployFunctions.lockUnlockFee = deployNewLockUnlockFee;
```
Comment (milapsheth) [inline test/InterchainTokenService.js:196]: :+1:

File: test/InterchainTokenService.js [hunk 8]
```diff
@@ -1325,6 +1338,59 @@ describe('Interchain Token Service', () => {
...
+                [sourceChain, destinationChain],
+                gateway,
+                gasService,
+                tokenManagerDeployer,
+                interchainTokenDeployer,
+                tokenHandler,
+            );
+        });
+
+        it(`Should revert on initiating an interchain token transfer for gateway token when gateway call failed`, async () => {
```
Comment (milapsheth) [inline test/InterchainTokenService.js:1361]:
```suggestion
        it('Should revert on initiating an interchain token transfer for gateway token when gateway call failed', async () => {
```

File: contracts/test/utils/TestGatewayCaller.sol (hunk 5 above)
Comment (ahramy) [reply contracts/test/utils/TestGatewayCaller.sol:8]: defined an error code `delegatecallFailed`. updated.

File: test/InterchainTokenService.js (hunk 6 above)
Comment (ahramy) [reply test/InterchainTokenService.js:32]: got it. updated

File: test/InterchainTokenService.js (hunk 7 above)
Comment (ahramy) [reply test/InterchainTokenService.js:196]: the code looks simpler and easier to understand. :)

File: test/InterchainTokenService.js (hunk 8 above)
Comment (ahramy) [reply test/InterchainTokenService.js:1361]: nice catch, updated.

File: contracts/test/utils/TestGatewayCaller.sol [hunk 9]
```diff
@@ -3,8 +3,9 @@
 pragma solidity ^0.8.0;
 
 contract TestGatewayCaller {
+    error delegatecallFailed();
```
Comment (milapsheth) [inline contracts/test/utils/TestGatewayCaller.sol:6]:
```suggestion
    error DelegatecallFailed();
```
we capitalize all error codes

**Review by milapsheth — APPROVED** [review]

File: contracts/test/utils/TestGatewayCaller.sol (hunk 9 above)
Comment (ahramy) [reply contracts/test/utils/TestGatewayCaller.sol:6]: I see. Updated.
